localstack/testing/testselection/matching.py

```diff
@@ -0,0 +1,170 @@
+
+
+def resolve_dependencies(module_name: str) -> set[str]:
+    """
+    Resolves dependencies for a given service module name
+
+    :param module_name: the name of the service to resolve (e.g. lambda_)
+    :return: set of resolved _service names_ that the service depends on (e.g. sts)
+    """
+    svc_name = _map_to_service_name(module_name)
+    return _expand_api_dependencies(svc_name)
+
+
+# TODO: might want to cache that, but for now it shouldn't be too much overhead
+def _reverse_dependency_map(dependency_map: dict[str, dict]) -> dict[str, set[str]]:
+    """
+    The current API_DEPENDENCIES actually maps the services to their own dependencies.
+    In our case here we need the inverse of this, we need to of which other services this service is a dependency of.
+    """
+    result = {}
+    for svc, deps in dependency_map.items():
+        for dep in deps:
+            result.setdefault(dep, set()).add(svc)
+    return result
+
+
+def _expand_api_dependencies(svc_name: str) -> set[str]:
+    result = set()
+
+    dependencies = _reverse_dependency_map(API_DEPENDENCIES).get(svc_name, [])
+    result.update(dependencies)
+
+    for dep in dependencies:
+        sub_deps = _expand_api_dependencies(dep)  # recursive call
+        result.update(sub_deps)
+    return result
+
+
+def get_test_dir_for_service(svc: str):
+    return f"tests/aws/services/{svc}"
+
+
+class Matcher:
+    def __init__(self, matching_func: Callable[[str], bool]):
+        self.matching_func = matching_func
+
+    def full_suite(self):
+        return lambda t: [SENTINEL_ALL_TESTS] if self.matching_func(t) else []
+
+    def ignore(self):
+        return lambda t: [SENTINEL_NO_TEST] if self.matching_func(t) else []
+
+    def service_tests(self, services: list[str]):
+        return (
+            lambda t: [get_test_dir_for_service(svc) for svc in services]
+            if self.matching_func(t)
+            else []
+        )
+
+    def passthrough(self):
+        return lambda t: [t] if self.matching_func(t) else []
+
+
+class Matchers:
+    @staticmethod
+    def glob(glob: str) -> Matcher:
+        return Matcher(lambda t: fnmatch.fnmatch(t, glob))
+
+    @staticmethod
+    def regex(glob: str) -> Matcher:
+        return Matcher(lambda t: bool(re.match(t, glob)))
+
+    @staticmethod
+    def prefix(prefix: str) -> Matcher:
+        return Matcher(lambda t: t.startswith(prefix))
+
+
+def generic_service_test_matching_rule(changed_file_path: str) -> set[str]:
+    """
+    Generic matching of changes in service files to their tests
+
+    :param changed_file_path: the file path of the detected change
+    :return: list of partial test file path filters for the matching service and all services it depends on
+    """
+    # TODO: consider API_COMPOSITES
+
+    match = re.findall("localstack/services/([^/]+)/.+", changed_file_path)
+    if not match:
+        match = re.findall(r"localstack/aws/api/([^/]+)/__init__\.py", changed_file_path)
```

Comment:
It feels a bit counter-intuitive that a small test file change would cause execution of the whole service tests, but I don't have a strong opinion on that and it might be a bit safer. 

We can maybe do an import analysis in the future that can provide an additional guide in detecting files, since sometimes test files also still import values, utils, etc. from other test files.

Reply:
Yeah that's true. We can also get back to this once we have some experience with writing the matchers. If it feels natural we can keep it, if it feel weird we can change it 😄

Reply:
Done. Let's go with including it for now and we can reconsider at a later point to make this more strict again. For now I agree its better to go with the safer approach that if in doubt, includes more tests.